san/impl/params_resolver.cpp

```diff
@@ -0,0 +1,123 @@
+#include <iostream>
+
+#include "san.hpp"
+namespace san {
+
+using namespace ast;
+
+void ParamsValidator::visit(Program* node) {
+    // initial value of a varible could be an expression
+    for (auto var : node->variables) {
+        if (var->type != nullptr) {
+            var->type->accept(*this);
+        }
```

Comment:
Just visit the variable declaration itself and let the appropriate overload handle visiting the type and the initial value